content/articles/integrate-ml-to-flask-api-and-deploy-to-heroku/index.md

```diff
@@ -0,0 +1,336 @@
+
+To build the Machine Learning model for this project, use this [guide](https://www.section.io/engineering-education/diagnose-diabetes-with-svm/). After training the model, we have to save it to a file. Serialization comes to play here. Serialization is the process of turning a Python object into a byte stream. The conventional method of serializing objects in Python is by using `pickle`. With `pickle`, we can serialize our machine learning model. We then save the serialized format to a file. It relieves us the stress of retraining our machine learning model each time we want to use it. Then, we can load the saved file and de-serialize it. Then, we use it to make new predictions.
+
+Add the code below to a new cell in your Jupyter Notebook. It pickles the machine learning model and training set. You may be wondering why we are pickling the training set. In this [guide](https://www.section.io/engineering-education/diagnose-diabetes-with-svm/), we feature scaled our test set with the mean and standard deviation of each feature in the training set. We need to use those same values to scale the new record before making predictions.
+
+```python
+# pickle the training set
+import pickle
+filename_X = 'X_train.sav'
+pickle.dump(X_train, open(filename_X, 'wb'))
+filename = 'model.sav'
+pickle.dump(model, open(filename, 'wb'))
+```
+
+Running the code above creates two new files in your current directory: `X_train.sav` and `model.sav`.
+
+
+### Integrating ML model to a Flask-RESTful API
+
+Using a text editor (VS Code), create a new file and rename it to `api.py.` 
+We create a virtual environment to isolate the libraries we will be using for this project. The virtual environment will not change the modules installed on our system. Instead, it will create a unique environment where we can install only the libraries we need for our project. Thus, it helps us avoid conflict between the libraries installed in our system and the libraries we will be using for our project. First, we install `virtualenv`. Then, we use `virtualenv` to build separate Python environments.
+
+```bash
+pip install virtualenv
+```
+
+We create a virtual environment and call it venv. It adds a new folder venv to our directory.
+
+```bash
+virtualenv venv
+```
+
+To activate the virtual environment on Windows, run the command below.
+
+```bash
+venv\Scripts\activate.bat 
+```
+
+Once we activate the virtual environment, we install all the libraries we need to run our API. The `pickle` module comes by default with python. So we do not need to install it. Next, we installed the version of the scikit-learn library we used to build the ML model in our Jupyter Notebook. To check for the version, run the code below in your notebook.
+
+```python
+import sklearn
+
+print('scikit-learn version is', sklearn.__version__)
+```
+
+**Output:**
+```
+scikit-learn version is 0.24.2
+```
+
+We install this version of scikit-learn in our virtual environment.
+
+```bash
+pip install flask

```

Comment:
Need to mention the exact versions used to avoid dependency errors in the future. If possible, also attach a requirements.txt file along with the tutorial and guide the reader to set up the environment accordingly.

Reply:
I listed all the libraries and their versions (in the requirements.txt) later in the article.